File: 0002-add-two-numbers/0002-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int carry = 0;
        ListNode head = new ListNode(-1), currNode = head;
        
        while(l1 != null || l2 != null){
            int l1Val = l1 == null ? 0 : l1.val; // 9
            int l2Val = l2 == null ? 0 : l2.val; // 9
            
            int currVal = l1Val + l2Val + carry; // 18
            currNode.next = new ListNode(currVal % 10); // 8
            carry = currVal / 10; // 1
            
            currNode = currNode.next;
            l1 = l1 == null ? null : l1.next;
            l2 = l2 == null ? null : l2.next;
        }
        
        if(carry != 0){
            currNode.next = new ListNode(carry);
        }
        
        return head.next;
    }
}

/*
[2,4,3]
[5,6,7,9,9,9]
[7,0,1,0,0,0,1]

l1 or 0 + l2 or 0 + carry
store mod in currNode 10 -> 0
carry = 1
next....

if carry != 0
    store carry

*/

File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        carry = 0
        head = ListNode(-1)
        dummy = ListNode()
        dummy.next = head
        
        while l1 or l2:
            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0
            
            currVal = val1 + val2 + carry
            carry = currVal / 10
            currVal = currVal % 10
            
            newNode = ListNode(currVal)
            head.next = newNode
            head = head.next
            
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
            
        if carry:
            head.next = ListNode(carry)
            
        return dummy.next.next
            
            
            
        